File: java/src/Solution0219.java
import java.util.HashSet;
import java.util.Set;

/**
 * Given an integer array nums and an integer k, return true if t
 * here are two distinct indices i and j in the array such that nums[i] == nums[j] and abs(i - j) <= k.
 *
 * Example 1:
 * Input: nums = [1,2,3,1], k = 3
 * Output: true
 *
 * Example 2:
 * Input: nums = [1,0,1,1], k = 1
 * Output: true

 * Example 3:
 * Input: nums = [1,2,3,1,2,3], k = 2
 * Output: false
 *
 * Constraints:
 * 1 <= nums.length <= 105
 * -109 <= nums[i] <= 109
 * 0 <= k <= 105
 *
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/contains-duplicate-ii
 * 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 */
public class Solution0219 {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        Set<Integer> set = new HashSet<>();
        for (int i = 0; i < nums.length; i++) {
            if (set.contains(nums[i])) {
                return true;
            }
            set.add(nums[i]);
            if (set.size() > k) {
                set.remove(nums[i - k]);
            }
        }
        return false;
    }
}

File: goR3/solution0219.go
package main

func ContainsNearbyDuplicate(nums []int, k int) bool {
	m := make(map[int][]int)

	for i := 0; i < len(nums); i++ {
		_, ok := m[nums[i]]
		if ok {
			t := m[nums[i]]
			if i-t[len(t)-1] <= k {
				return true
			} else {
				t = append(t, i)
				m[nums[i]] = t
			}
		} else {
			t := []int{i}
			m[nums[i]] = t
		}
	}

	return false
}
